python-avd/pyavd/api/pool_manager/base_classes.py

```diff
@@ -0,0 +1,282 @@
+# Copyright (c) 2023-2025 Arista Networks, Inc.
+# Use of this source code is governed by the Apache License 2.0
+# that can be found in the LICENSE file.
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import asdict, dataclass, field
+from typing import TYPE_CHECKING, Generic, TypeVar, cast
+
+from yaml import CSafeDumper, CSafeLoader, dump, load
+
+from pyavd.j2filters import natural_sort
+
+if TYPE_CHECKING:
+    from pathlib import Path
+
+    from typing_extensions import Self
+
+    from pyavd._eos_designs.shared_utils import SharedUtils
+
+T_ValueType = TypeVar("T_ValueType", int, str)
+T_PoolKeyType = TypeVar("T_PoolKeyType", bound="PoolKey")
+T_AssignmentKeyType = TypeVar("T_AssignmentKeyType", bound="AssignmentKey")
+
+FILE_HEADER = """\
+# This file is auto-generated by AVD eos_designs.
+# When there is a merge conflict for this file, rebase the branch, accept all current changes for this file, and re-run AVD.
+"""
+
+
+@dataclass(frozen=True)
+class PoolKey:
+    @classmethod
+    def from_dict(cls, data: dict) -> Self:
+        return cls(**data)
+
+
+@dataclass(frozen=True)
+class AssignmentKey:
+    @classmethod
+    def from_dict(cls, data: dict) -> Self:
+        return cls(**data)
+
+
+@dataclass
+class PoolAssignment(Generic[T_AssignmentKeyType, T_ValueType]):
+    key: T_AssignmentKeyType
+    value: T_ValueType
+    accessed: bool = False
+
+    def as_dict(self) -> dict:
+        return {"key": asdict(self.key), "value": self.value}
+
+
+@dataclass
+class Pool(Generic[T_PoolKeyType, T_AssignmentKeyType, T_ValueType]):
+    """One Pool of resources. Currently only supporting a pool of Integers."""
+
+    collection: PoolCollection[T_PoolKeyType, T_AssignmentKeyType, T_ValueType]
+    pool_key: T_PoolKeyType
+    assignments: dict[T_AssignmentKeyType, PoolAssignment[T_AssignmentKeyType, T_ValueType]]
+
+    def _is_value_available(self, value: T_ValueType) -> bool:
+        if self.collection.value_type is int and isinstance(value, int):
+            collection = cast(PoolCollection[T_PoolKeyType, T_AssignmentKeyType, int], self.collection)
+            assignments = cast(dict[T_AssignmentKeyType, PoolAssignment[T_AssignmentKeyType, int]], self.assignments)
```

Comment:
why do we need to cast?

Reply:
I don't think this worked in py3.10 type hinting. After upgrading to 3.12 I don't need this.